Sources/SourceKitLSP/Rename.swift

```diff
@@ -284,78 +291,438 @@ private extension RenameLocation.Usage {
-      // First, group all occurrences of that USR by the files they occur in.
-      var locationsByFile: [URL: [RenameLocation]] = [:]
-      let occurrences = index.occurrences(ofUSR: usr, roles: [.declaration, .definition, .reference])
-      for occurrence in occurrences {
-        let url = URL(fileURLWithPath: occurrence.location.path)
-        let renameLocation = RenameLocation(
-          line: occurrence.location.line,
-          utf8Column: occurrence.location.utf8Column,
-          usage: RenameLocation.Usage(roles: occurrence.roles)
-        )
-        locationsByFile[url, default: []].append(renameLocation)
-      }
+    let renameResult = try await primaryFileLanguageService.rename(request)
 
-      // Now, call `editsToRename(locations:in:oldName:newName:)` on the language service to convert these ranges into
-      // edits.
-      let urisAndEdits =
-        await locationsByFile
-        .filter { changes[DocumentURI($0.key)] == nil }
-        .concurrentMap { (url: URL, renameLocations: [RenameLocation]) -> (DocumentURI, [TextEdit])? in
-          let uri = DocumentURI(url)
-          // Create a document snapshot to operate on. If the document is open, load it from the document manager,
-          // otherwise conjure one from the file on disk. We need the file in memory to perform UTF-8 to UTF-16 column
-          // conversions.
-          // We should technically infer the language for the from-disk snapshot. But `editsToRename` doesn't care
-          // about it, so defaulting to Swift is good enough for now
-          // If we fail to get edits for one file, log an error and continue but don't fail rename completely.
-          guard
-            let snapshot = (try? self.documentManager.latestSnapshot(uri))
-              ?? (try? DocumentSnapshot(url, language: .swift))
-          else {
-            logger.error("Failed to get document snapshot for \(uri.forLogging)")
-            return nil
-          }
-          do {
-            let edits = try await languageService.editsToRename(
-              locations: renameLocations,
-              in: snapshot,
-              oldName: oldName,
-              newName: request.newName
-            )
-            return (uri, edits)
-          } catch {
-            logger.error("Failed to get edits for \(uri.forLogging): \(error.forLogging)")
+    guard let usr = renameResult.usr, let index = workspace.index else {
+      // We don't have enough information to perform a cross-file rename.
+      return renameResult.edits
+    }
+
+    guard
+      let (oldTranslatableName, nativeLanguage) = try await getTranslatableName(
+        forUsr: usr,
+        workspace: workspace,
+        index: index
+      )
+    else {
+      // We failed to get the translatable name, so we can't to global rename.
+      // Do local rename within the current file instead as fallback.
+      return renameResult.edits
+    }
```

Comment:
Unfortunate that we redo this after the prepare. But... theoretically the AST is in the cache and that would be the most expensive part. Plus I assume it's not *required* to send a prepare first?

Reply:
Exactly, the prepare is really just a “Can you rename this? And if so, what’s the symbol’s range and what name should I pre-populate the input field with?”. I don’t think it’s worth adding another layer of caching for this.